Adding AGENTS.md (#4573)
kptdev/kpt: AGENTS.md

# Cloud Agent Instructions for kpt

For Copilot, Cursor, Codex, Gemini Code Assist, or any agents generating code or reviewing pull requests in this repository.

## Generic rules

### Coding

* Never commit directly to this repository, always aim for a pull request
* Always ask for a final review from a human before committing
* Make sure that AI agent usage is attributed in the correct way

### Reviewing PRs

> * When committing suggested changes, add a sign-off from the approving human and add the `Assisted-by` tag to the commit message

Reviewer: This is confusing on who needs to be the author and who needs be in the `Assisted-by` annotation

Author: Fixed in [eb6c0fe](https://github.com/kptdev/kpt/pull/4573/commits/eb6c0fe5beb1393b4e0a810a50b35bdbbf4d619b)

- Go (version is specified in [go.mod](./go.mod))
- Git (required and checked at runtime)
- Docker or Podman (for `test-docker` and function runtime tests)
- KinD (CI uses the version defined in `.github/workflows/e2eEnvironment.yml`)

### Build

```bash
make build
```
This compiles kpt to `$(go env GOPATH)/bin/kpt` using LDFLAGS with git commit SHA.

### Run All Checks (Build, Test, Lint)

```bash
make all
```

Runs: `fix vet fmt lint test build tidy` in sequence. *Always* run this before committing.

### Unit Tests

```bash
make test
```

Runs Go tests with coverage. Set `KRM_FN_RUNTIME` to select function runtime (docker/podman, default uses system default).

### Docker/Podman Runtime Tests

```bash
make test-docker
```

Requires Docker or Podman. Tests that need container runtime (e.g., pipeline tests). Respects `KRM_FN_RUNTIME` environment variable.

### E2E Function Tests

#### Render tests

```bash
make test-fn-render T=".*"
```
#### Eval tests

```bash
make test-fn-eval T=".*"
```

Use `T` parameter to filter tests by regex (e.g., `T=fnconfig` for function config tests). Set `KRM_FN_RUNTIME` to select runtime.

### E2E Live Apply Tests

```bash
make test-live-apply
```

> Requires KinD with Kubernetes 1.33.4 and 1.34.0 (specific SHAs pinned in CI). These tests use Kind internally. Timeout is 20 minutes.

Reviewer 2: ```suggestion
Requires KinD with Kubernetes 1.35.5 and 1.36.1 (specific SHAs pinned in CI). These tests use Kind internally. Timeout is 20 minutes.
```

Reviewer 3: is there a better way of doing this so we don't have to update the values every time? something to check latest stable kind version & associated k8s version supported by that kind version?

Author: I've made an attempt to fix this in [96783d2](https://github.com/kptdev/kpt/pull/4573/commits/96783d2d015527c5ef5163797c05e97aaa991bbd). There are only references to the locations where the versions are actually defined.

* **Makefile**: Primary build orchestration
* **.golangci.yml**: Linting configuration (used golangci-lint version is defined in the `GOLANGCI_LINT_VERSION` variable of the [Makefile](./Makefile))
* **go.mod / go.sum**: Dependency management
* **CONTRIBUTING.md**: Contribution guidelines and code review requirements
* **CODEOWNERS**: Default reviewers

### Key Source Directories

* **commands/**: CLI command implementations using Cobra framework
* **run/**: Main CLI setup; contains GetMain() that initializes Cobra commands with environment setup
* **pkg/**: Core library packages (business logic, utilities)
* **internal/**: Internal packages; includes internal/docs/generated/ (generated from Markdown via mdtogo)
* **mdtogo/**: Code generator tool that converts CLI documentation Markdown files to Go variables

### Build & CI Configuration

* **.github/workflows/go.yml**: Main CI workflow
  * Runs on Linux (docker/podman matrix) and macOS
  * Executes: `make all` + `make test-docker`
  * Triggered on PRs (except changes limited to ignored paths like `**.md`) and pushes (pushes exclude dependabot branches)
* **.github/workflows/e2eEnvironment.yml**: KinD-based e2e tests

>   * Tests Kubernetes 1.33 and 1.34 with KinD v0.30.0

Reviewer 2: ```suggestion
  * Tests Kubernetes 1.35 and 1.36 with KinD v0.32.0
```

Author: Fixed in [96783d2](https://github.com/kptdev/kpt/pull/4573/commits/96783d2d015527c5ef5163797c05e97aaa991bbd)

* **.github/workflows/live-e2e.yml**: Live apply e2e tests
  * Tests with pinned Kubernetes image SHAs
  * Runs `make test-live-apply` with `K8S_VERSION` environment variable
* **.github/workflows/release-api.yml**: Release the kpt API module
  * Builds, tests and lints the kpt API module
  * Runs `make api`

### Documentation & Tests

* **documentation/**: Hugo-based website published to kpt.dev
  * Run `make serve` from `documentation/` (serves docs locally)
  * Requires `npm install` first (run in `documentation/`)

* **e2e/**: End-to-end test suites
  * Contains testdata directories for function render/eval tests
  * Live tests in `e2e/live/end-to-end-test.sh`

### Other Notable Directories

* **release/**: Release automation (GoReleaser config, Homebrew formula generation)
* **hack/**: Miscellaneous development utilities
* **healthcheck/**: Separate module for health checking (Go (Go version is defined in [healthcheck/go.mod](./healthcheck/go.mod)), local Makefile)
* **thirdparty/**: Third-party code (excluded from linting)
* **Formula/**: Homebrew package definition (generated by `go run ./release/formula/main.go VERSION`)

> ## Linting Rules & Style

Reviewer 3: should we be linking this to the kpt docs guide https://github.com/kptdev/kpt/tree/main/documentation#style-guide-for-documentation or is this just for code linting?

Author: This is only for code linting, but I've added a reference to the style guide to the docs section in [96783d2](https://github.com/kptdev/kpt/pull/4573/commits/96783d2d015527c5ef5163797c05e97aaa991bbd).

### Linter Configuration

* Linter configuration is defined in the [.golangci.yml](./.golangci.yml) file

### Code Style Requirements (from CONTRIBUTING.md)

* **Copyright Headers**: All files must have Apache 2.0 license header
  * Use format: // Copyright YEAR The kpt Authors (or year range if modified)

>   * Year should match creation year, or creation-to-modification year range

Reviewer 2: It would be good to point out the detailed copyright year update instruction available on porch documentation https://porch.kpt.dev/docs/12_contributing/code-contribution/#update-copyright-on-files"

Author: Fixed in [96783d2](https://github.com/kptdev/kpt/pull/4573/commits/96783d2d015527c5ef5163797c05e97aaa991bbd).

* **Developer Certificate of Origin (DCO)**: Commits must be signed with -s flag
* **Large Features**: Require reviewed and merged design document (use /docs/design-docs/00-template.md as template)
* **AI Usage**: Must declare AI usage in the PR description; `Assisted-by: AGENT_NAME:MODEL_VERSION` attribution in
  commit messages is recommended (see CONTRIBUTING.md).

## Validation Checklist for PRs

Before submitting a PR, verify:

1. ✅ All tests pass: make all
1. ✅ All linting passes: make lint
1. ✅ Code formatted: make fmt
1. ✅ Dependencies tidied: make tidy
1. ✅ Copyright headers added/updated per CONTRIBUTING.md
1. ✅ DCO sign-off: use git commit -s
1. ✅ For CLI/API changes: design document reviewed and merged
1. ✅ AI usage declared in PR description (if applicable)

## Common Build Issues & Workarounds

### Docker/Podman Not Available

If `make test-docker` fails due to missing Docker/Podman:

* Install Docker Desktop or Podman
* For Podman: ensure it's on PATH and `podman version` runs successfully
* Set `KRM_FN_RUNTIME=podman` if using Podman

### KinD Setup Issues

For e2e live tests (`make test-live-apply`):

> * KinD v0.30.0 is auto-installed by CI workflow

Reviewer 2: ```suggestion
* KinD v0.32.0 is auto-installed by CI workflow
```

Author: Fixed in [96783d2](https://github.com/kptdev/kpt/pull/4573/commits/96783d2d015527c5ef5163797c05e97aaa991bbd)

* Requires Docker running in background
* Tests use specific Kubernetes image SHAs (see live-e2e.yml matrix)
* Timeout is 20 minutes; allow sufficient time

### Git Configuration

Tests fail silently if git user not configured. Always run:

```bash
git config --global user.email "you@example.com"
git config --global user.name "Your Name"
```

### Module Generation

If changes modify CLI documentation in documentation/content/en/reference/cli/:

* Run `make generate` to regenerate `internal/docs/generated/`
* Commit generated files

### Known CI Skips

> * Windows build currently disabled (see `.github/workflows/go.yml` line 88-104, issue #3463)

Reviewer 2: ```suggestion
* Windows build currently disabled (see `.github/workflows/go.yml`, issue #3463)
```
This can be removed as this gets updated on every change in that file.

### Run All Checks (Build, Test, Lint)

```bash
make all
```

Runs: `fix vet fmt lint test build tidy` in sequence. *Always* run this before committing.

### Unit Tests

```bash
make test
```

Runs Go tests with coverage. Set `KRM_FN_RUNTIME` to select function runtime (docker/podman, default uses system default).

### Docker/Podman Runtime Tests

```bash
make test-docker
```

Requires Docker or Podman. Tests that need container runtime (e.g., pipeline tests). Respects `KRM_FN_RUNTIME` environment variable.

### E2E Function Tests

#### Render tests

```bash
make test-fn-render T=".*"
```
#### Eval tests

```bash
make test-fn-eval T=".*"
```

Use `T` parameter to filter tests by regex (e.g., `T=fnconfig` for function config tests). Set `KRM_FN_RUNTIME` to select runtime.

### E2E Live Apply Tests

```bash
make test-live-apply
```

Requires KinD with Kubernetes 1.33.4 and 1.34.0 (specific SHAs pinned in CI). These tests use Kind internally. Timeout is 20 minutes.

### Linting

```bash
make lint
```

Runs golangci-lint v2.11.4. If already installed locally with matching version, uses it; otherwise downloads and runs via `go run`.

### Code Generation & Formatting

```bash

> make fmt        # Run gofmt and goimports

Author: Fixed in [fdeaee6](https://github.com/kptdev/kpt/pull/4573/commits/fdeaee650aa83db6f2c2f14773ca6cdc9688f260)

  * Run `make serve` from `documentation/` (serves docs locally)
  * Requires `npm install` first (run in `documentation/`)

* **e2e/**: End-to-end test suites
  * Contains testdata directories for function render/eval tests
  * Live tests in `e2e/live/end-to-end-test.sh`

### Other Notable Directories

* **release/**: Release automation (GoReleaser config, Homebrew formula generation)
* **hack/**: Miscellaneous development utilities
* **healthcheck/**: Separate module for health checking (Go 1.26.3, local Makefile)
* **thirdparty/**: Third-party code (excluded from linting)
* **Formula/**: Homebrew package definition (generated by `go run ./release/formula/main.go VERSION`)

## Linting Rules & Style

### Linter Configuration

* **Enabled Linters**: bodyclose, copyloopvar, dogsled, dupl, errcheck, gochecknoinits, goconst, gocritic, gocyclo, govet, ineffassign, lll, misspell, nakedret, revive, staticcheck, unconvert, unparam, unused, whitespace
* **Duplication Threshold**: 400 lines
* **Cyclomatic Complexity**: max 30
* **Line Length**: max 170 characters
* **Revive Confidence**: 0.85
* **Excluded Paths**: thirdparty/, third_party, builtin, examples

> * **Test Files**: Further relaxed rules (gosec, funlen disabled for *_test.go files)

Author: Fixed in [a2f8a33](https://github.com/kptdev/kpt/pull/4573/commits/a2f8a331e499f819fc1c80fb0fc4f0def6a7be3b)

### Code Style Requirements (from CONTRIBUTING.md)

* **Copyright Headers**: All files must have Apache 2.0 license header
  * Use format: // Copyright YEAR The kpt Authors (or year range if modified)

>    * Year should match creation year, or creation-to-modification year range

Author: Fixed in [a2f8a33](https://github.com/kptdev/kpt/pull/4573/commits/a2f8a331e499f819fc1c80fb0fc4f0def6a7be3b)

* **Developer Certificate of Origin (DCO)**: Commits must be signed with -s flag
* **Large Features**: Require reviewed and merged design document (use /docs/design-docs/00-template.md as template)

> * **AI Usage**: Must declare AI usage in PR description and commit messages with `Assisted-by: AGENT_NAME:MODEL_VERSION` format

Author: Fixed in [a2f8a33](https://github.com/kptdev/kpt/pull/4573/commits/a2f8a331e499f819fc1c80fb0fc4f0def6a7be3b)

# Cloud Agent Instructions for kpt

For Copilot, Cursor, Codex, Gemini Code Assist or any agents generating code or review Pull Requests
in this repository. 

## Generic rules

### Coding

* Never commit directly to this repository, always aim for a pull request
* Always ask for a final review of a human before commiting
* Make sure that AI agent usage is attributed in the correct way

### Reviewing PRs

* When comitting suggested changes add a signoff of the approving human and add the `Assisted-by`

>   tag to the commit message

Author: In case of AI agents committing changes the signoff should happen by the person who approves the changes.

## Repository Overview

**kpt** is a package-centric toolchain that automates Kubernetes configuration editing and management. It enables declarative configuration authoring, automation, and delivery at scale through a "Configuration as Data" approach, supporting Kubernetes platforms and KRM-driven infrastructure (e.g., Config Connector, Crossplane).

> - **Language**: Go 1.26.3

Reviewer 4: Should the version be included?

Author: Correct I will just refer to the `go.mod`

Author: Fixed in [e9072e6](https://github.com/kptdev/kpt/pull/4573/commits/e9072e6dc09a42698eb1bd314e643295d9ebb097)

> - **Repository Size**: ~78 MB

Reviewer 4: I don't think this is meaningful info like this, we could put an approx line or file count, but that would also have to be kept up to date.

Author: Yep, I will remove this.

Author: Fixed in [e9072e6](https://github.com/kptdev/kpt/pull/4573/commits/e9072e6dc09a42698eb1bd314e643295d9ebb097)

- **License**: Apache 2.0
- **Key Topics**: Kubernetes, configuration management, CLI tooling, GitOps, policy-as-code, KRM (Kubernetes Resource Model)

## Quick Build & Test Commands

**Trust these instructions first.** Only perform searches if you find this information incomplete or inaccurate.

### Prerequisites

> - Go 1.26.3 (specified in `go.mod`)

Reviewer 4: I think we should instruct the AI to check the go.mod for this so we don't have to update it here too.

Author: Fixed in [e9072e6](https://github.com/kptdev/kpt/pull/4573/commits/e9072e6dc09a42698eb1bd314e643295d9ebb097)

- Git (required and checked at runtime)
- Docker or Podman (for `test-docker` and function runtime tests)

> - KinD (CI uses v0.30.0 for live e2e/apply workflows; `make install-kind` installs v0.29.0 and is used by `verifyContent.yml`) (for e2e live apply tests with Kubernetes 1.33 and 1.34)

Reviewer 4: Again, we should point it at the file containing the version, not specify it straight up.

Author: Fixed in [e9072e6](https://github.com/kptdev/kpt/pull/4573/commits/e9072e6dc09a42698eb1bd314e643295d9ebb097)

### Run All Checks (Build, Test, Lint)

```bash
make all
```

Runs: `fix vet fmt lint test build tidy` in sequence. *Always* run this before committing.

### Unit Tests

```bash
make test
```

Runs Go tests with coverage. Set `KRM_FN_RUNTIME` to select function runtime (docker/podman, default uses system default).

### Docker/Podman Runtime Tests

```bash
make test-docker
```

Requires Docker or Podman. Tests that need container runtime (e.g., pipeline tests). Respects `KRM_FN_RUNTIME` environment variable.

### E2E Function Tests

#### Render tests

```bash
make test-fn-render T=".*"
```
#### Eval tests

```bash
make test-fn-eval T=".*"
```

Use `T` parameter to filter tests by regex (e.g., `T=fnconfig` for function config tests). Set `KRM_FN_RUNTIME` to select runtime.

### E2E Live Apply Tests

```bash
make test-live-apply
```

Requires KinD with Kubernetes 1.33.4 and 1.34.0 (specific SHAs pinned in CI). These tests use Kind internally. Timeout is 20 minutes.

### Linting

```bash
make lint
```

Runs golangci-lint v2.11.4. If already installed locally with matching version, uses it; otherwise downloads and runs via `go run`.

### Code Generation & Formatting

```bash

> make fmt        # Run gofmt and goimports

Author: Fixed

make tidy       # Run go mod tidy

> make generate   # Generate code from templates (mdtogo, copyright headers)

Reviewer 4: I think generate runs schema generation too, please check

Author: Nope, it just populates `internal/docs/generated` with mdtogo

make schema     # Generate schema
```

### Setting Up Git Config (Required for Tests)

Many tests require git configuration:

```bash
git config --global user.email "you@example.com"
git config --global user.name "Your Name"
```

## Project Layout

### Root Level Files & Directories

* **main.go**: CLI entry point; contains //go:generate directives for CLI documentation
* **Makefile**: Primary build orchestration
* **.golangci.yml**: Linting configuration (golangci-lint v2.11.4)

> * **go.mod / go.sum**: Dependency management (Go 1.26.3)

Reviewer 4: Again, I think we should remove the versions.

Author: Fixed in [e9072e6](https://github.com/kptdev/kpt/pull/4573/commits/e9072e6dc09a42698eb1bd314e643295d9ebb097)

* **.golangci.yml**: Linting configuration (golangci-lint v2.11.4)
* **go.mod / go.sum**: Dependency management (Go 1.26.3)
* **CONTRIBUTING.md**: Contribution guidelines and code review requirements
* **CODEOWNERS**: Default reviewers

### Key Source Directories

* **commands/**: CLI command implementations using Cobra framework
* **run/**: Main CLI setup; contains GetMain() that initializes Cobra commands with environment setup
* **pkg/**: Core library packages (business logic, utilities)
* **internal/**: Internal packages; includes internal/docs/generated/ (generated from Markdown via mdtogo)
* **mdtogo/**: Code generator tool that converts CLI documentation Markdown files to Go variables

### Build & CI Configuration

* **.github/workflows/go.yml**: Main CI workflow
  * Runs on Linux (docker/podman matrix) and macOS
  * Executes: `make all` + `make test-docker`

>   * Triggered on PRs and pushes (excludes dependabot branches)

Author: Fixed in [a2f8a33](https://github.com/kptdev/kpt/pull/4573/commits/a2f8a331e499f819fc1c80fb0fc4f0def6a7be3b)

* **.github/workflows/verifyContent.yml**: Verifies CLI examples
  * Runs `make build`, installs mdrip/kind, runs `make site-verify-examples`

>   * Triggered on changes to `commands/`, `internal/` paths

Author: There is no site directory. I will create a separate pr to fix this.

>   * Triggered on changes to `commands/`, `internal/` paths

Author: The whole `.github/workflows/verifyContent.yml` workflow will be removed in https://github.com/kptdev/kpt/issues/4574

>   * Triggered on changes to `commands/`, `internal/` paths

Author: Removed from here also in [f1a88ef](https://github.com/kptdev/kpt/pull/4573/commits/f1a88ef44b4fd2b9fb14e707658ef9b3cdc6b3ba)

  * Runs `./e2e/live/end-to-end-test.sh -k <K8S_VERSION>`
* **.github/workflows/live-e2e.yml**: Live apply e2e tests
  * Tests with pinned Kubernetes image SHAs

>   * Runs `make test-live-apply` with `K8S_VERSION` environment variable

Reviewer 4: Be sure to update this if the API module separation goes in before this PR

Author: https://github.com/kptdev/kpt/pull/4537 was just merged :)

Author: Added in [e9072e6](https://github.com/kptdev/kpt/pull/4573/commits/e9072e6dc09a42698eb1bd314e643295d9ebb097)

### Documentation & Tests

* **documentation/**: Hugo-based website published to kpt.dev
  * Run `make serve` from `documentation/` (serves docs locally)
  * Requires `npm install` first (run in `documentation/`)

* **e2e/**: End-to-end test suites
  * Contains testdata directories for function render/eval tests
  * Live tests in `e2e/live/end-to-end-test.sh`

### Other Notable Directories

* **release/**: Release automation (GoReleaser config, Homebrew formula generation)
* **hack/**: Miscellaneous development utilities

> * **healthcheck/**: Separate module for health checking (Go 1.26.3, local Makefile)

Reviewer 4: Another go version ref

Author: Fixed in [e9072e6](https://github.com/kptdev/kpt/pull/4573/commits/e9072e6dc09a42698eb1bd314e643295d9ebb097)

* **thirdparty/**: Third-party code (excluded from linting)
* **Formula/**: Homebrew package definition (generated by `go run ./release/formula/main.go VERSION`)

## Linting Rules & Style

### Linter Configuration

* **Enabled Linters**: bodyclose, copyloopvar, dogsled, dupl, errcheck, gochecknoinits, goconst, gocritic, gocyclo, govet, ineffassign, lll, misspell, nakedret, revive, staticcheck, unconvert, unparam, unused, whitespace
* **Duplication Threshold**: 400 lines
* **Cyclomatic Complexity**: max 30
* **Line Length**: max 170 characters
* **Revive Confidence**: 0.85
* **Excluded Paths**: thirdparty/, third_party, builtin, examples
* **Test Files**: `funlen` and `gosec` are currently disabled project-wide (see `.golangci.yml`); if enabled later,

>   test files have additional exclusions configured.

Reviewer 4: Again, I think we should point it at the file so we don't have to keep this info up to date here.

Author: Fixed in [e9072e6](https://github.com/kptdev/kpt/pull/4573/commits/e9072e6dc09a42698eb1bd314e643295d9ebb097)

## Common Build Issues & Workarounds

### Docker/Podman Not Available

If `make test-docker` fails due to missing Docker/Podman:

* Install Docker Desktop or Podman
* For Podman: ensure it's on PATH and `podman version` runs successfully
* Set `KRM_FN_RUNTIME=podman` if using Podman

### KinD Setup Issues

For e2e live tests (`make test-live-apply`):

* KinD v0.30.0 is auto-installed by CI workflow
* Requires Docker running in background
* Tests use specific Kubernetes image SHAs (see live-e2e.yml matrix)
* Timeout is 20 minutes; allow sufficient time

### Git Configuration

Tests fail silently if git user not configured. Always run:

```bash
git config --global user.email "you@example.com"
git config --global user.name "Your Name"
```

### Module Generation

If changes modify CLI documentation in documentation/content/en/reference/cli/:

* Run `make generate` to regenerate `internal/docs/generated/`
* Commit generated files

### Known CI Skips

* Windows build currently disabled (see `.github/workflows/go.yml` line 88-104, issue #3463)
* Some linters disabled: `funlen`, `gosec` (marked TODO in `.golangci.yml`)

## Environment Variables

> * **KRM_FN_RUNTIME**: Select function runtime for tests: `docker`,  `podman` or `nerdctl`

Author: Fixed in [f1a88ef](https://github.com/kptdev/kpt/pull/4573/commits/f1a88ef44b4fd2b9fb14e707658ef9b3cdc6b3ba)
